### ISSUE TITLE
CoreWebView2CreateCoreWebView2ControllerCompletedHandler never invoked in DLL

### ISSUE THREAD

**Issue by carlsound** [MicrosoftEdge/WebView2Feedback#920]
I am trying to use WebView2 for a Win32 DLL for a plug-in.

The ICoreWebView2Environment->CreateCoreWebView2Controller(HWND, CoreWebView2CreateCoreWebView2ControllerCompletedHandler) method appears to invoke the CoreWebView2CreateCoreWebView2ControllerCompletedHandler callback in the getting started EXE, but not in a DLL using similar code.

Are there additional steps I need to perform in the case of a DLL?
For instance, I did learn that CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED) needed to be included in the case of a DLL, before calling the CreateCoreWebView2Environment() function...

**Comment by LiangTheDev** [comment]
We have to also pump Windows message on the thread where we try to create WebView2. The related code in the [sample app](https://github.com/MicrosoftEdge/WebView2Samples/blob/master/SampleApps/WebView2APISample/App.cpp) is the `while (GetMessage(&msg, nullptr, 0, 0))` loop.

**Comment by carlsound** [comment]
Yes, the DLL I am creating has the Windows message pump on the same thread where the WebView2 instance is created.

**Comment by carlsound** [comment]
ICoreWebView2Environment->CreateCoreWebView2Controller() returns S_OK

the ICoreWebView2CreateCoreWebView2ControllerCompletedHandler is not invoked, even after waiting ~2 minutes

no Microsoft Edge WebView2 process is created after CreateCoreWebView2Controller is called

**Comment by LiangTheDev** [comment]
Is it possible to share the related WebView2 creation code?
We don't expect there is any difference whether the caller code is in a dll or exe, and expect the completed handler always to be called.

**Comment by carlsound** [comment]
Call to DLL:
`
m_webView2LoaderDll = LoadLibraryA("WebView2Loader.dll");
if (m_webView2LoaderDll)
{
     m_controllerviewDll = LoadLibraryA("Win32WebView2Test_ControllerView.dll");
     if (!m_controllerviewDll)
     {
          m_dllCreateChildWindowFunction = (void (*) (HWND)) GetProcAddress(m_controllerviewDll, "createWindow");
          if (m_dllCreateChildWindowFunction)
          {
	           m_dllCreateChildWindowFunction(m_parentWindow);
          }
     }
}
`

DLL createWindow function:
`
void createWindow(HWND parentHwnd)
{
    CreateThread(NULL, 0, (LPTHREAD_START_ROUTINE)UI_Thread, (void*)parentHwnd, 0, 0);
}
`

DLL child window UI Thread:
`
DWORD WINAPI UI_Thread(LPVOID lpParam)
{
     HWND parentWindow = (HWND*)lpParam;
     RECT  parentWindowRect{ 0 };
     if (parentWindow) { GetWindowRect(parentWindow, &parentWindowRect); }
      //
		WNDCLASSEXW windowClass = { 0 };
		windowClass.cbSize = sizeof(WNDCLASSEX);
		windowClass.cbClsExtra = 10;
		windowClass.cbWndExtra = 10;
		windowClass.hCursor = LoadCursor(0, IDC_ARROW);
		windowClass.hIcon = LoadIcon(0, IDI_WINLOGO);
		windowClass.lpszMenuName = NULL;
		windowClass.style = CS_HREDRAW | CS_VREDRAW;;
		windowClass.hbrBackground = 0;
		windowClass.lpfnWndProc = WndProc;
		windowClass.hInstance = ghInst; //m_hinstance;
		windowClass.lpszClassName = TEXT("WebView2");
		//
		ATOM windowClassRegistration = RegisterClassEx(&windowClass);
		//
		if (windowClassRegistration)
		{
                      		HWND childWindow = CreateWindowEx(
				windowClass.style,			 
				windowClass.lpszClassName,     
				windowClass.lpszClassName,    
				WS_VISIBLE | WS_CHILD,           
				0,			
				0,			
				parentWindowRect.right - parentWindowRect.left, 
				parentWindowRect.bottom - parentWindowRect.top,
				parentWindow,                  
				nullptr,                         
				windowClass.hInstance,         
				NULL); 
                        if (NULL != childWindow)
			{
				ShowWindow(childWindow, SW_SHOWDEFAULT); //SW_SHOW);
				UpdateWindow(childWindow);
                                CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);
                                HRESULT hr1 = CreateCoreWebView2Environment
				(
					Microsoft::WRL::Callback<ICoreWebView2CreateCoreWebView2EnvironmentCompletedHandler>
					(
						//this,
						&OnEnvironmentCreatedAsync
						).Get()
				);
                                MSG msg;
                               while (GetMessage(&msg, childWindow, 0, 0))
			       {
					TranslateMessage(&msg);
				        DispatchMessage(&msg);
			         }
                                 CoUninitialize();
				return (int)msg.wParam;
			}
                }
        return 0;
}
`

DLL WebView2 create environment function:
`
	HRESULT OnEnvironmentCreatedAsync(
		HRESULT result,
		ICoreWebView2Environment* environment)
	{
		if (environment)
		{
			HRESULT hr2 = environment->CreateCoreWebView2Controller
			(
				webViewWindow,
				Microsoft::WRL::Callback<ICoreWebView2CreateCoreWebView2ControllerCompletedHandler>
				(
					&OnControllerCreatedAsync
				).Get() 
			);
		}
                return result;
        }
`

DLL WebView2 create controller function:
`
	HRESULT OnControllerCreatedAsync(
		HRESULT errorCode,
		ICoreWebView2Controller* createdController)
	{
		if !createdController)
		{
			createdController->get_CoreWebView2(&webviewWindow);
		}
		return errorCode;
	}
`

**Comment by LiangTheDev** [comment]
Try change
`while (GetMessage(&msg, childWindow, 0, 0))`
to
`while (GetMessage(&msg, nullptr, 0, 0))`

nullptr means pump message for all window on the thread, not just the childWindow. WebView creates other window and need to pump message for those window as well.

Another thing, what type is webViewWindow? We are doing:
`environment->CreateCoreWebView2Controller ( webViewWindow,`
and later
`if !createdController) { createdController->get_CoreWebView2(&webviewWindow);`
We need provide hwnd when calling CreateCoreWebView2Controller.
And also I assume that it is `if (createdController)`.

**Comment by carlsound** [comment]
while (GetMessage(&msg, nullptr, 0, 0)) {...}
resolved the issue.
Thank you so much!
![Screenshot 2021-02-18 225832](https://user-images.githubusercontent.com/12228422/108460366-23420480-723e-11eb-9abb-b4c2e51019b9.png)

**Comment by MikeYeager** [comment]
We're having a similar problem. We've got a C++ wrapper around the WebView2 (evergreen) and are calling it from a .NET application. The first time we create a new window with a WebView2, it works perfectly. When we try to create a second window with a WebView2 (in a separate process), it returns S_OK, but the completed handler never fires. It does fire eventually after the first window is closed. Unfortunately, the code is a bit verbose. We've got MessageBoxes in there for debugging. On the second instance, we get these MessageBoxes: "This is B!", "Entering AttachWebView", "QI 1A", "QI 1B", then nothing... The original programmer put the flag in there to to make the initialization synchronous, but from our testing, that doesn't seem to play into the problem. We put a 5 second timeout in the event loop to ensure the flag isn't preventing it from exiting the loop if it doesn't work. Is there anyway to determine WHY the completed handler does not fire? Any logging or some way to determine what's preventing it from firing? Thank you in advance.

`void Photino::AttachWebView()
{
	std::atomic_flag flag = ATOMIC_FLAG_INIT;
	flag.test_and_set();

	MessageBox(_hWnd, L"Entering AttachWebView", L"Debug", MB_OK);
	HRESULT envResult = CreateCoreWebView2EnvironmentWithOptions(nullptr, nullptr, nullptr,

		Callback<ICoreWebView2CreateCoreWebView2EnvironmentCompletedHandler>(
			[&](HRESULT result, ICoreWebView2Environment* env) -> HRESULT { 

			CHECK_FAILURE(result);
			if (result != S_OK) { return result; }

			MessageBox(_hWnd, L"QueryInterface1A", L"Debug", MB_OK);
			CHECK_FAILURE(env->QueryInterface(&_webviewEnvironment));
			MessageBox(_hWnd, L"QueryInterface1B", L"Debug", MB_OK);

			std::wostringstream s;
			s << std::hex << &_webviewEnvironment;
			std::wstring strTitle = s.str();
			MessageBox(_hWnd, strTitle.c_str(), L"Debug", MB_OK);

			// Create a WebView, whose parent is the main window hWnd
			HRESULT mike = env->CreateCoreWebView2Controller(_hWnd, 

				Callback<ICoreWebView2CreateCoreWebView2ControllerCompletedHandler>(
					[&](HRESULT result, ICoreWebView2Controller* controller) -> HRESULT {

					MessageBox(_hWnd, L"ICoreWebView2CreateCoreWebView2ControllerCompletedHandler", L"Debug", MB_OK);

					if (controller != nullptr)
						_webviewController = controller;
					else
						MessageBox(NULL, L"Controller is null", L"", MB_OK);

					CHECK_FAILURE(result);
					if (result != S_OK) { return result; }

					MessageBox(_hWnd, L"QueryInterface2A", L"Debug", MB_OK);
					//HRESULT envResult = controller->QueryInterface(&_webviewController);
					MessageBox(_hWnd, L"QueryInterface2B", L"Debug", MB_OK);

					if (envResult != S_OK) { return envResult; }

					_webviewController->get_CoreWebView2(&_webviewWindow);

					// Add a few settings for the webview
					// this is a redundant demo step as they are the default settings values
					ICoreWebView2Settings* Settings;
					_webviewWindow->get_Settings(&Settings);
					Settings->put_IsScriptEnabled(TRUE);
					Settings->put_AreDefaultScriptDialogsEnabled(TRUE);
					Settings->put_IsWebMessageEnabled(TRUE);

					// Register interop APIs
					EventRegistrationToken webMessageToken;
					_webviewWindow->AddScriptToExecuteOnDocumentCreated(L"window.external = { sendMessage: function(message) { window.chrome.webview.postMessage(message); }, receiveMessage: function(callback) { window.chrome.webview.addEventListener(\'message\', function(e) { callback(e.data); }); } };", nullptr);
					_webviewWindow->add_WebMessageReceived(

						Callback<ICoreWebView2WebMessageReceivedEventHandler>(
						[&](ICoreWebView2* webview, ICoreWebView2WebMessageReceivedEventArgs* args) -> HRESULT {
						
							MessageBox(_hWnd, L"This is B!", L"Debug", MB_OK);
							wil::unique_cotaskmem_string message;
							args->TryGetWebMessageAsString(&message);
							_webMessageReceivedCallback(message.get());
							return S_OK;
						}).Get(), &webMessageToken);

					EventRegistrationToken webResourceRequestedToken;
					_webviewWindow->AddWebResourceRequestedFilter(L"*", COREWEBVIEW2_WEB_RESOURCE_CONTEXT_ALL);
					_webviewWindow->add_WebResourceRequested(
						
						Callback<ICoreWebView2WebResourceRequestedEventHandler>(
						[&](ICoreWebView2* sender, ICoreWebView2WebResourceRequestedEventArgs* args) {
							MessageBox(_hWnd, L"C", L"Debug", MB_OK);		ICoreWebView2WebResourceRequest* req;
							args->get_Request(&req);

							wil::unique_cotaskmem_string uri;
							req->get_Uri(&uri);
							std::wstring uriString = uri.get();
							size_t colonPos = uriString.find(L':', 0);
							if (colonPos > 0)
							{
								std::wstring scheme = uriString.substr(0, colonPos);
								WebResourceRequestedCallback handler = _schemeToRequestHandler[scheme];
								if (handler != NULL)
								{
									int numBytes;
									AutoString contentType;
									wil::unique_cotaskmem dotNetResponse(handler(uriString.c_str(), &numBytes, &contentType));

									if (dotNetResponse != nullptr && contentType != nullptr)
									{
										std::wstring contentTypeWS = contentType;

										IStream* dataStream = SHCreateMemStream((BYTE*)dotNetResponse.get(), numBytes);
										wil::com_ptr<ICoreWebView2WebResourceResponse> response;
										_webviewEnvironment->CreateWebResourceResponse(
											dataStream, 200, L"OK", (L"Content-Type: " + contentTypeWS).c_str(),
											&response);
										args->put_Response(response.get());
									}
								}
							}

							return S_OK;
						}
					).Get(), &webResourceRequestedToken);

					MessageBox(_hWnd, L"D", L"Debug", MB_OK);
					RefitContent();

					flag.clear();
					return S_OK;
				}).Get());

			if (mike == S_OK)
				MessageBox(NULL, L"OK", L"FFFF", MB_OK);
			else
				MessageBox(NULL, L"NOT OK", L"FFFF", MB_OK);
			//flag.clear();
			return S_OK;
		}).Get());

	CHECK_FAILURE(envResult);

	if (envResult != S_OK)
	{
		_com_error err(envResult);
		LPCTSTR errMsg = err.ErrorMessage();
		MessageBox(_hWnd, errMsg, L"Error instantiating webview", MB_OK);
	}
	else
	{
		MessageBox(NULL, L"Begin Blocking", L"", MB_OK);
		// Block until it's ready. This simplifies things for the caller, so they don't need to regard this process as async.
		MSG msg = { };
		int counter = 0;
		while (flag.test_and_set() && GetMessage(&msg, nullptr, 0, 0))
		{
			if (counter++ > 200) break;
			TranslateMessage(&msg);
			DispatchMessage(&msg);
		}
		MessageBox(NULL, L"I Quit", L"", MB_OK);
	}
}`

**Comment by LiangTheDev** [comment]
If we are calling AttachWebView() from _webMessageReceivedCallback, I suspend that this is the issue of nested message loop, which WebView2 code currently doesn't support.
When WebMessageReceived is fired and _webMessageReceivedCallback runs, we are inside DispatchMessage of the message loop for the first WebView2, when we call AttachWebView() from there, we are entering the message loop again. This nested message loop could happen naturally when show some dialog window, and WebView2 code would not process any further event in this nested message loop to avoid various re-entering issues.
For our usage of creating a second WebView2, there is no need to run a second message loop. We only need one message loop per thread. If we could avoid the second message loop, then it should work.

**Comment by MikeYeager** [comment]
@LiangTheDev Thank you for your reply. If I understand correctly, you're saying that when there is more than one instance of the native window, they all share the same message loop? There is one message loop per thread, not per instance? 

FYI: I created a simplified demo of this with full source code in #1164

**Comment by LiangTheDev** [comment]
That is correct. We only need one message loop per thread, not per instance. The one message loop will pump messages for all instances.

**Comment by MikeYeager** [comment]
@LiangTheDev You sir, are a rock star! It's working now. One thing you guys might want to add to the docs is to navigate to a page in the controller completed callback handler. That eliminates the need to block when initializing a WebView2 and then load the initial page in most cases (since the user can't navigate again until the initial page is loaded). Several of us have attempted to block until initialization is finished so that it doesn't blow up when navigating to that first page. In fact, it would be nice if the initial page to navigate to could be passed down through the callbacks. A multi-window sample we recently found, navigates to "bing.com", a fixed string. In our case, we added the initial page string to our constructor and set an instance variable on our wrapper class. A little bit of a hack as the memory is not released, but it does allow us to reference it from the callback code. It's not as clean as passing it down into the callback. THANK YOU AGAIN!
